File: client/src/models/UserSettings.ts
import {
  Model, attribute as attr, initialize, STRING, BOOLEAN, NUMBER,
} from './Model'

export enum DailyReminderType {
  TEXT = 'text',
  EMAIL = 'email',
  NONE = 'none',
}

export class UserSettings extends Model {
  static modelName = 'userSettings'

  constructor(data?: any) {
    super()
    initialize(this, data)
  }

  @attr({ type: BOOLEAN })
  overrideLocationReminders = false

  @attr({ type: STRING })
  dailyReminderType: DailyReminderType = DailyReminderType.TEXT

  @attr({ type: NUMBER })
  dailyReminderTime = 7

  @attr({ type: BOOLEAN })
  remindMon = false

  @attr({ type: BOOLEAN })
  remindTue = false

  @attr({ type: BOOLEAN })
  remindWed = false

  @attr({ type: BOOLEAN })
  remindThu = false

  @attr({ type: BOOLEAN })
  remindFri = false

  @attr({ type: BOOLEAN })
  remindSat = false

  @attr({ type: BOOLEAN })
  remindSun = false

  dailyReminderHour(): number {
    return this.dailyReminderTime > 12 ? this.dailyReminderTime - 12 : this.dailyReminderTime
  }

  dailyReminderAMPM(): 'am' | 'pm' {
    if (this.dailyReminderTime === 24) return 'am'
    return this.dailyReminderTime < 12 && this.dailyReminderTime !== 12 ? 'am' : 'pm'
  }
}

File: app/models/user_settings.rb
# frozen_string_literal: true

class UserSettings < ApplicationRecord
  extend Enumerize

  DAILY_REMINDER_TYPES = [
    TEXT = 'text',
    EMAIL = 'email',
    NONE = 'none'
  ].freeze

  enumerize :daily_reminder_type, in: DAILY_REMINDER_TYPES

  belongs_to :user

  before_create :match_reminder_type_to_user_contact

  def match_reminder_type_to_user_contact
    if user.mobile_number?
      TEXT
    elsif user.email?
      EMAIL
    else
      NONE
    end
  end
end

# == Schema Information
#
# Table name: user_settings
#
#  id                          :bigint           not null, primary key
#  user_id                     :bigint           not null
#  override_location_reminders :boolean          default(FALSE), not null
#  daily_reminder_type         :string           default("text"), not null
#  daily_reminder_time         :integer          default(7), not null
#  remind_mon                  :boolean          default(TRUE), not null
#  remind_tue                  :boolean          default(TRUE), not null
#  remind_wed                  :boolean          default(TRUE), not null
#  remind_thu                  :boolean          default(TRUE), not null
#  remind_fri                  :boolean          default(TRUE), not null
#  remind_sat                  :boolean          default(TRUE), not null
#  remind_sun                  :boolean          default(TRUE), not null
#  created_at                  :datetime         not null
#  updated_at                  :datetime         not null
#
# Indexes
#
#  index_user_settings_on_user_id  (user_id)
#
# Foreign Keys
#
#  fk_rails_...  (user_id => users.id) ON DELETE => cascade
#
